File: app/src/main/java/com/example/pocedex/presentation/PokemonCardActivity.java
package com.example.pocedex.presentation;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Toast;

import com.example.pocedex.R;
import com.example.pocedex.data.CommentAndFavorite;
import com.example.pocedex.domain.LocalSave;
import com.example.pocedex.domain.Network;
import com.example.pocedex.data.Pokemon;
import com.example.pocedex.databinding.ActivityPokeCardBinding;
import com.example.pocedex.domain.Utils;

import java.util.ArrayList;
import java.util.List;

public class PokemonCardActivity extends AppCompatActivity implements IUpdatePokemon {

    private static String pokemonlink = "";
    CommentAndFavorite cardCommentAndFavorite;
    Pokemon pokemon;
    ActivityPokeCardBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (!Utils.isOnline(this)) {
            setContentView(R.layout.offline);
            return;
        }
        setContentView(R.layout.activity_poke_card);
        Bundle arguments = getIntent().getExtras();
        pokemonlink = arguments.get("link").toString();
        new Network().getPokemon(this, pokemonlink, this);
    }

    @Override
    public void refresh(List<Pokemon> pokemons)
    {
        setPokemon(pokemons.get(0));
    }

    public void setPokemon(Pokemon newPokemon)
    {
        pokemon=newPokemon;
        binding = DataBindingUtil.setContentView(this, R.layout.activity_poke_card);

        binding.setPokemon(pokemon);
        cardCommentAndFavorite = findOnId(pokemon.getId(), LocalSave.getCommentAndFavorites());
        if (cardCommentAndFavorite == null) {
            cardCommentAndFavorite = new CommentAndFavorite();
            cardCommentAndFavorite.setId(pokemon.getId());
            cardCommentAndFavorite.setName(pokemon.getName());
            cardCommentAndFavorite.setUrl(pokemonlink);
        }
        if (cardCommentAndFavorite.getIsFav()) {
            binding.favBtn.setImageResource(android.R.drawable.star_big_on);
        }
        if (cardCommentAndFavorite.getComment() != null) {
            binding.UsCom.setText(cardCommentAndFavorite.getComment());
        }
        if (pokemon.getSprite(5) == null) {
            binding.sp5.setVisibility(View.GONE);
            binding.sp6.setVisibility(View.GONE);
            binding.sp7.setVisibility(View.GONE);
            binding.sp8.setVisibility(View.GONE);
        }
    }

    public CommentAndFavorite findOnId(
            int id, ArrayList<CommentAndFavorite> caf) {
        try {
            for (CommentAndFavorite c : caf) {
                if (c.equals(id)) {
                    return c;
                }
            }
        } catch (Exception e) {
            Log.d("", e.getMessage());
        }
        return null;
    }

    public void saveComm(View view) {
        cardCommentAndFavorite.setComment(binding.UsCom.getText().toString());
        //cardCommentAndFavorite.setComment(((EditText) findViewById(R.id.UsCom)).getText().toString());
        showToast("Your comment save");
        save();
        hideKeyboard(this);
    }

    public static void hideKeyboard(Activity activity) {
        InputMethodManager imm = (InputMethodManager) activity.getSystemService(Activity.INPUT_METHOD_SERVICE);
        View view = activity.getCurrentFocus();
        if (view == null) {
            view = new View(activity);
        }
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }

    public void saveFav(View view) {
        cardCommentAndFavorite.setIsFav(!cardCommentAndFavorite.getIsFav());
        if (cardCommentAndFavorite.getIsFav()) {
            binding.favBtn.setImageResource(android.R.drawable.star_big_on);
            showToast("Save in Fav");
        } else {
            binding.favBtn.setImageResource(android.R.drawable.star_big_off);
            showToast("Delete from Fav");
        }
        save();
    }

    private void save() {
        if (findOnId(pokemon.getId(), LocalSave.getCommentAndFavorites()) == null) {
            LocalSave.addToCommentAndFavorites(cardCommentAndFavorite);
        }
        try {
            LocalSave.save();
        } catch (Exception e) {
            Log.d("comfavsave", e.getMessage());
        }
    }

    private void showToast(String mes) {
        Toast toast = Toast.makeText(this, mes, Toast.LENGTH_LONG);
        toast.show();
    }
}

File: app/src/main/java/com/example/pocedex/presentation/PokemonCardActivity.kt
package com.example.pocedex.presentation

import android.app.Activity
import android.app.Dialog
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.InsetDrawable
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.ImageButton
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.pocedex.R
import com.example.pocedex.data.CommentAndFavorite
import com.example.pocedex.data.Pokemon
import com.example.pocedex.databinding.ActivityCardOriginBinding
import com.example.pocedex.domain.Network
import com.example.pocedex.domain.Utils
import com.google.gson.Gson
import java.util.ArrayList

internal class PokemonCardActivity : AppCompatActivity(), IUpdatePokemon, INetworkChange {

    private var pokemonlink = ""
    private var cardCommentAndFavorite: CommentAndFavorite? = null
    private var pokemon: Pokemon = Pokemon()
    private lateinit var bindingOrig: ActivityCardOriginBinding
    private var edit: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bindingOrig = DataBindingUtil.setContentView(this, R.layout.activity_card_origin)
        val arguments = intent.extras
        val gson = Gson()
        pokemon = gson.fromJson(arguments!!.get("pokemon")!!.toString(), Pokemon::class.java)
        pokemonlink = pokemon.url.toString()
        if (pokemon.getSprite(0) != null)
            setPokemon(pokemon)
        else {
            if (Utils.isConnected) {
                setOnline()
            } else {
                setOffline()
            }
        }
    }

    override fun onStop() {
        super.onStop()
        Utils.stopNetworkCallback(this)
    }

    override fun onStart() {
        super.onStart()
        Utils.startNetworkCallback(this)
    }

    private fun setOffline() {
        bindingOrig.llayOffline.visibility = View.VISIBLE
        bindingOrig.llayCard.visibility = View.GONE
    }

    override fun setOnline() {
        if (Utils.isConnected) {
            bindingOrig.llayCard.visibility = View.VISIBLE
            bindingOrig.llayOffline.visibility = View.GONE
            Network().getPokemon(this, pokemonlink, this)
        }
    }


    override fun refresh(pokemons: List<Pokemon>) {
        setPokemon(pokemons[0])
    }

    private fun setPokemon(newPokemon: Pokemon) {
        pokemon = newPokemon

        bindingOrig.pokemon = pokemon
        cardCommentAndFavorite = Utils.findOnId(pokemon.safeId, Utils.localSave!!.getCommentAndFavorites())
        if (cardCommentAndFavorite == null) {
            cardCommentAndFavorite = CommentAndFavorite()
            cardCommentAndFavorite!!.pokemon = pokemon
            cardCommentAndFavorite!!.pokemon.url = pokemonlink
        }
        if (cardCommentAndFavorite!!.is_favorite) {
            bindingOrig.incCard.btnFavorite.setImageResource(android.R.drawable.star_big_on)
        }
        if (!cardCommentAndFavorite!!.comment?.isEmpty()!!) {
            bindingOrig.incCard.etxtComment.visibility = View.VISIBLE
            bindingOrig.incCard.etxtComment.setText(cardCommentAndFavorite!!.comment)
        }
    }

    fun showSprite(view: View) {
        if (view.id != R.id.imv_pokemon_image)
            return
        val spriteDialog = Dialog(this)
        spriteDialog.setContentView(R.layout.sprite_list)
        val back = ColorDrawable(Color.TRANSPARENT)
        val inset = InsetDrawable(back, 40)
        val recyclerView = spriteDialog.findViewById<RecyclerView>(R.id.rv_pokemons_list)
        val sprites: ArrayList<String> = ArrayList()
        for (i in 1..8) {
            if (pokemon.getSprite(i) != null)
                sprites.add(pokemon.getSprite(i)!!)
        }
        val adapter = SpriteAdapter(sprites)
        recyclerView.layoutManager = LinearLayoutManager(spriteDialog.context, LinearLayoutManager.HORIZONTAL, false)
        recyclerView.adapter = adapter

        spriteDialog.window!!.setBackgroundDrawable(inset)
        spriteDialog.show()
    }

    fun editComment(view: View) {
        edit = !edit
        if (edit) {
            bindingOrig.incCard.etxtComment.visibility = View.VISIBLE
            bindingOrig.incCard.etxtComment.isEnabled = true
            (view as ImageButton).setImageResource(android.R.drawable.ic_menu_save)
        } else {
            (view as ImageButton).setImageResource(android.R.drawable.ic_menu_edit)
            bindingOrig.incCard.etxtComment.isEnabled = false
            if (bindingOrig.incCard.etxtComment.text.isEmpty())
                bindingOrig.incCard.etxtComment.visibility = View.GONE
            else
                saveComm()
        }
    }

    override fun repeat() {}

    private fun saveComm() {
        cardCommentAndFavorite!!.comment = bindingOrig.incCard.etxtComment.text.toString()
        showToast(R.string.save_comment)
        Utils.save(pokemon, cardCommentAndFavorite!!)
        hideKeyboard(this)
    }

    private fun hideKeyboard(activity: Activity) {
        val imm = activity.getSystemService(Activity.INPUT_METHOD_SERVICE) as InputMethodManager
        var view = activity.currentFocus
        if (view == null) {
            view = View(activity)
        }
        imm.hideSoftInputFromWindow(view.windowToken, 0)
    }

    fun saveFav(view: View) {
        if (view.id != R.id.btn_favorite)
            return
        cardCommentAndFavorite!!.is_favorite = (!cardCommentAndFavorite!!.is_favorite)
        if (cardCommentAndFavorite!!.is_favorite) {
            bindingOrig.incCard.btnFavorite.setImageResource(android.R.drawable.star_big_on)
            showToast(R.string.save_to_fav)
        } else {
            bindingOrig.incCard.btnFavorite.setImageResource(android.R.drawable.star_big_off)
            showToast(R.string.delete_from_fav)
        }
        Utils.save(pokemon, cardCommentAndFavorite!!)
    }

    private fun showToast(mes: Int) {
        val toast = Toast.makeText(this, mes, Toast.LENGTH_LONG)
        toast.show()
    }
}